radicle-httpd/src/api/v1/projects.rs

```diff
@@ -0,0 +1,392 @@
+    Path((project, sha, path)): Path<(Id, Oid, String)>,
+) -> impl IntoResponse {
+    let path = path.strip_prefix('/').ok_or(Error::NotFound)?.to_string();
+    let storage = &ctx.profile.storage;
+    let repo = storage.repository(project)?;
+    let tree = radicle_surf::object::tree(&repo.raw().into(), Some(Sha { sha }), Some(path))?;
+    let response = json!({
+        "path": &tree.path,
+        "entries": &tree.entries,
+        "info": &tree.info,
+        "stats": stats(&repo)?,
+    });
+
+    Ok::<_, Error>(Json(response))
+}
+
+/// Get all project remotes.
+/// `GET /projects/:project/remotes`
+async fn remotes_handler(
+    Extension(ctx): Extension<Context>,
+    Path(project): Path<Id>,
+) -> impl IntoResponse {
+    let storage = &ctx.profile.storage;
+    let repo = storage.repository(project)?;
+    let remotes = repo
+        .remotes()?
+        .filter_map(|r| r.map(|r| r.1).ok())
+        .collect::<Vec<_>>();
+
+    Ok::<_, Error>(Json(remotes))
+}
+
+/// Get project remote.
+/// `GET /projects/:project/remotes/:peer`
+async fn remote_handler(
+    Extension(ctx): Extension<Context>,
+    Path((project, node_id)): Path<(Id, NodeId)>,
+) -> impl IntoResponse {
+    let storage = &ctx.profile.storage;
+    let repo = storage.repository(project)?;
+    let remote = repo
+        .remotes()?
+        .filter(|r| {
+            if let Ok((id, _)) = r {
+                node_id == *id
+            } else {
+                false
+            }
+        })
+        .filter_map(|r| r.map(|r| r.1).ok())
+        .collect::<Vec<_>>();
+
+    Ok::<_, Error>(Json(remote))
+}
+
+/// Get project source file.
+/// `GET /projects/:project/blob/:sha/*path`
+async fn blob_handler(
+    Extension(ctx): Extension<Context>,
+    Path((project, sha, path)): Path<(Id, Oid, String)>,
+) -> impl IntoResponse {
+    let path = path.strip_prefix('/').ok_or(Error::NotFound)?;
+    let storage = &ctx.profile.storage;
+    let repo = storage.repository(project)?;
+    let blob = radicle_surf::blob::blob(&repo.raw().into(), Some(Sha { sha }), path)?;
+
+    Ok::<_, Error>(Json(blob))
+}
+
+/// Get project readme.
+/// `GET /projects/:project/readme/:sha`
+async fn readme_handler(
+    Extension(ctx): Extension<Context>,
+    Path((project, sha)): Path<(Id, Oid)>,
+) -> impl IntoResponse {
+    let storage = &ctx.profile.storage;
+    let repo = storage.repository(project)?;
+    let paths = &[
+        "README",
+        "README.md",
+        "README.markdown",
+        "README.txt",
+        "README.rst",
+        "Readme.md",
+    ];
+
+    for path in paths {
+        if let Ok(blob) = radicle_surf::blob::blob(&repo.raw().into(), Some(Sha { sha }), path) {
+            return Ok::<_, Error>(Json(blob));
+        }
+    }
+
+    Err(radicle_surf::object::Error::PathNotFound(
+        radicle_surf::file_system::Path::try_from("README").unwrap(),
+    ))?
+}
+
+/// Get project patches list.
+/// `GET /projects/:project/patches`
```

Comment:
Are these all patches whether open and closed?

A future suggestion is to also paginate these.

Reply:
this is all for now, should we add pagination in this PR as well @cloudhead ?

Reply:
Let's keep that for another PR :+1: